ext/web/16_report_error.js

```diff
@@ -0,0 +1,76 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+"use strict";
+
+((window) => {
+  const core = window.Deno.core;
+  const { ErrorEvent } = window;
+  const {
+    Error,
+    PromiseResolve,
+    PromisePrototypeThen,
+    StringPrototypeStartsWith,
+  } = window.__bootstrap.primordials;
+
+  let printException = undefined;
+
+  /** Set a function `(e: Error) => void` to print an uncaught exception. It is
+  invoked when an error event is dispatched by the runtime and
+  `Event.prototype.preventDefault()` isn't called. */
+  function setPrintException(fn) {
+    printException = fn;
+  }
+
+  let reportExceptionStackedCalls = 0;
+
+  // https://html.spec.whatwg.org/#report-the-exception
+  function reportException(error) {
+    reportExceptionStackedCalls++;
+    const jsError = core.destructureError(error);
+    const message = jsError.message;
+    let filename = "";
+    let lineno = 0;
+    let colno = 0;
+    if (jsError.frames.length > 0) {
+      filename = jsError.frames[0].fileName;
+      lineno = jsError.frames[0].lineNumber;
+      colno = jsError.frames[0].columnNumber;
+    } else {
+      const jsError = core.destructureError(new Error());
+      for (const frame of jsError.frames) {
+        if (
+          typeof frame.fileName == "string" &&
+          !StringPrototypeStartsWith(frame.fileName, "deno:")
+        ) {
+          filename = frame.fileName;
+          lineno = frame.lineNumber;
+          colno = frame.columnNumber;
+          break;
+        }
+      }
+    }
+    const event = new ErrorEvent("error", {
+      cancelable: true,
+      message,
+      filename,
+      lineno,
+      colno,
+      error,
+    });
+    // Avoid recursing `reportException()` via error handlers more than once.
+    if (reportExceptionStackedCalls > 1 || window.dispatchEvent(event)) {
+      printException?.(jsError);
+      // TODO(nayeemrmn): Use `queueMicrotask()` instead once it's fixed
+      // (https://github.com/denoland/deno/issues/14158). Largely because the
+      // `(in promise)` isn't desired.
+      PromisePrototypeThen(PromiseResolve(), () => {
+        throw new Error(`Unhandled error event.`);
+      });
```

Comment:
Discussed with @lucacasonato and we don't really get what's going on here. 

It feels like we should not call `printException` and just directly call `throw error`.

See:
```js
// foo.js
globalThis.addEventListener("error", (e) => {
    console.log("foo");
    e.preventDefault();
})

setTimeout(() => {
    throw new Error("boom!")
}, 1000);
```

```
❯ cargo run -- run foo.js
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
     Running `target/debug/deno run foo.js`
foo2

deno on  error-event-for-events-and-timers [$!?] via  v17.4.0 via 🦀 v1.59.0 took 2s
❯ echo $status
0

```

Removing `e.preventDefault()` from the event listener yields:
```
❯ cargo run -- run foo.js
   Compiling deno_runtime v0.53.0 (/Users/ib/dev/deno/runtime)
   Compiling deno v1.20.4 (/Users/ib/dev/deno/cli)
    Finished dev [unoptimized + debuginfo] target(s) in 23.51s
     Running `target/debug/deno run foo.js`
foo2
error: Uncaught Error: boom!
    throw new Error("boom!")
          ^
    at file:///Users/ib/dev/deno/foo.js:10:11
    at Object.action (deno:ext/web/02_timers.js:145:13)
    at handleTimerMacrotask (deno:ext/web/02_timers.js:62:12)

deno on  error-event-for-events-and-timers [$!?] via  v17.4.0 via 🦀 v1.59.0 took 26s
❯ echo $status
1

```

Please explain the reasoning for current solution

Reply:
> It feels like we should not call `printException` and just directly call `throw error`.

It's so the error is printed synchronously as it is in browsers, and the actual error leading to termination is detached. I especially wanted this for `reportError()`.

Reply:
> It's so the error is printed synchronously as it is in browsers

Wouldn't `throw error` also achieve this?

> and the actual error leading to termination is detached

Why is this important? It seems rather annoying to get two errors printed if something goes wrong.

Reply:
> Wouldn't `throw error` also achieve this?

Without a microtask that makes it catchable around `dispatchEvent(new CustomEvent("leads-to-error"))` and `reportError()` which is undesired. The other option is throwing the main error in the microtask, should I just do that?

Reply:
Ah I see... I think we might want to tackle this differently. What about adding a core binding in the form `core.terminate(error)` that saves the error onto the isolate state, then triggers an isolate termination ([`Isolate::terminate_execution`](https://docs.rs/v8/latest/v8/struct.Isolate.html#method.terminate_execution)), which is essentially an uncatchable error. Then once this termination bubbles up to the place in `deno_core` where we'd return an error back to the Rust caller we grab the error from the isolate state and return that. Does that make sense?